src/Umbraco.Web/Models/Mapping/ContentVariantMapper.cs

```diff
@@ -21,52 +21,110 @@ public ContentVariantMapper(ILocalizationService localizationService)
 
         public IEnumerable<ContentVariantDisplay> Map(IContent source, MapperContext context)
         {
-            var result = new List<ContentVariantDisplay>();
-            if (!source.ContentType.VariesByCulture())
+            var variesByCulture = source.ContentType.VariesByCulture();
+            var variesBySegment = source.ContentType.VariesBySegment();
+
+            IList<ContentVariantDisplay> variants = new List<ContentVariantDisplay>();
+
+            if (!variesByCulture && !variesBySegment)
             {
-                //this is invariant so just map the IContent instance to ContentVariationDisplay
-                result.Add(context.Map<ContentVariantDisplay>(source));
+                // this is invariant so just map the IContent instance to ContentVariationDisplay
+                var variantDisplay = context.Map<ContentVariantDisplay>(source);
+                variants.Add(variantDisplay);
+            }
+            else if (variesByCulture && !variesBySegment)
+            {
+                var languages = GetLanguages(context);
+                variants = languages
+                    .Select(language => CreateVariantDisplay(context, source, language, null))
+                    .ToList();
+            }
+            else if (variesBySegment && !variesByCulture)
+            {
+                // Segment only
+                var segments = GetSegments(source);
+                variants = segments
+                    .Select(segment => CreateVariantDisplay(context, source, null, segment))
+                    .ToList();
             }
             else
             {
-                var allLanguages = _localizationService.GetAllLanguages().OrderBy(x => x.Id).ToList();
-                if (allLanguages.Count == 0) return Enumerable.Empty<ContentVariantDisplay>(); //this should never happen
-
-                var langs = context.MapEnumerable<ILanguage, Language>(allLanguages).ToList();
+                // Culture and segment
+                var languages = GetLanguages(context).ToList();
+                var segments = GetSegments(source).ToList();
 
-                //create a variant for each language, then we'll populate the values
-                var variants = langs.Select(x =>
+                if (languages.Count == 0 || segments.Count == 0)
                 {
-                    //We need to set the culture in the mapping context since this is needed to ensure that the correct property values
-                    //are resolved during the mapping
-                    context.SetCulture(x.IsoCode);
-                    return context.Map<ContentVariantDisplay>(source);
-                }).ToList();
+                    // This should not happen
+                    throw new ArgumentException("No languages or segments available");
```

Comment:
Probably InvalidOperationException here, don't think ArgumentException is the right one

Reply:
Agree, pushed the change